pkg/roachprod/vm/gce/dns.go

```diff
@@ -23,34 +23,46 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
 	"github.com/cockroachdb/errors"
 	"golang.org/x/exp/maps"
-	"golang.org/x/sync/errgroup"
 )
 
 const (
-	dnsManagedZone           = "roachprod-managed"
-	dnsDomain                = "roachprod-managed.crdb.io"
-	dnsMaxResults            = 10000
-	dnsMaxConcurrentRequests = 4
+	dnsManagedZone = "roachprod-managed"
+	dnsDomain      = "roachprod-managed.crdb.io"
+	dnsMaxResults  = 10000
 )
 
 var ErrDNSOperation = fmt.Errorf("error during Google Cloud DNS operation")
 
 var _ vm.DNSProvider = &dnsProvider{}
 
+type ExecFn func(cmd *exec.Cmd) ([]byte, error)
+
 // dnsProvider implements the vm.DNSProvider interface.
 type dnsProvider struct {
 	recordsCache struct {
 		mu      syncutil.Mutex
 		records map[string][]vm.DNSRecord
 	}
+	execFn ExecFn
 }
 
 func NewDNSProvider() vm.DNSProvider {
+	var gcloudMu syncutil.Mutex
+	return NewDNSProviderWithExec(func(cmd *exec.Cmd) ([]byte, error) {
+		// Limit to one gcloud command at a time.
```

Comment:
I wonder to what extent this will help. We have several release branches running at the same time, so we will continue to make concurrent `gcloud` calls across those builds.

Reply:
Yes, this is not for global exclusion, but is meant for local exclusion:

The unknown for me here is if `gcloud` is *locally* okay with multiple interactions with the same resource. In the sense that I'm concerned it's creating some kind of local cache of the object (by name). And then a second `gcloud` process has possible race issues.

I have nothing concrete to base it on though, since technically `gcloud` CLI isn't open source. The dual calls to the update the same entry that was present in the logs made me think it could be locally caching somewhere.

Reply:
I'm however concerned this might add some drag to the test suites, especially if many tests are starting services at the same time they would have to "queue". Thoughts? This is a bit of a knee-jerk reaction to the rate-limit we hit, but I still don't have a clear explanations for the double API calls we could see `gcloud` API making - I think I'll create a support ticket to ask about it.

Reply:
> In the sense that I'm concerned it's creating some kind of local 'file-based' cache of the object (by name).

Ah I see. If we decide to keep this code, it's probably worth to document this intention, as it'll be hard for someone to understand why we're doing this unless they trace it back to this discussion.

> I have nothing concrete to base it on though, since technically gcloud CLI isn't open source.

The source control instance is not public, but the source code is available in the `gcloud` releases themselves (it's Python). There are also some unofficial mirrors you can find on GitHub that can make it easier to navigate or clone the source. See [this one](https://github.com/google-cloud-sdk-unofficial/google-cloud-sdk), for example.

> I'm however concerned this might add some drag to the test suites, especially if many tests are starting services at the same time they would have to "queue". Thoughts?

I think the best way to determine is to run a nightly build with and without this change and compare the running time (a `0.1` build should be sufficient to measure the difference since that's still over 100 tests). I have a feeling it won't add significant overhead, but better safe than sorry.

Provided there's no local shenanigans in the `gcloud` client, we could perhaps have a more clear notion of [rate limiting](https://pkg.go.dev/golang.org/x/time/rate) instead of serializing every call.

> I think I'll create a support ticket to ask about it.

I agree; we should understand why that was (is) happening.